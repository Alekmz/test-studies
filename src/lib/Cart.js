import find from "lodash/find";
import remove from "lodash/remove";

export default class Cart {
  items = [];

  add(item) {
    const itemToFindOrRemove = { product: item.product };
    if (find(this.items, itemToFindOrRemove)) {
      remove(this.items, itemToFindOrRemove);
    }
    this.items.push(item);
  }

  remove(product) {
    remove(this.items, { product });
  }

  getTotal() {
    return this.items.reduce((acc, item) => {
      return acc + item.quantity * item.product.price;
    }, 0);
  }

  sumary() {
    const total = this.getTotal();
    const items = this.items;

    return {
      total,
      items,
    };
  }

  checkout() {
    const { total, items } = this.sumary();

    this.items = [];
    return {
      total,
      items,
    };
  }
}
