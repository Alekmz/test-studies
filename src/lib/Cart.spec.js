import Cart from "./Cart";

describe("Cart test", () => {
  let cart;

  let product = {
    title: "Nike shoes",
    price: 45054,
  };

  let product2 = {
    title: "Adidas shoes",
    price: 55060,
  };

  beforeEach(() => {
    cart = new Cart();
  });

  describe("getTotal()", () => {
    it("should return 0 when getTotal() is executed in a newly created instance", () => {
      expect(cart.getTotal()).toEqual(0);
    });

    it("should multiply the quantity of products by the price and return the total", () => {
      const itemProduct = {
        product,
        quantity: 2, // total 90108
      };

      cart.add(itemProduct);

      expect(cart.getTotal()).toEqual(90108);
    });

    it("should ensure no more than on prodcut exists at a time", () => {
      cart.add({
        product,
        quantity: 2,
      });
      cart.add({
        product,
        quantity: 2,
      });

      expect(cart.getTotal()).toEqual(90108);
    });

    it("should update total when a product gets indluded and then removed", () => {
      cart.add({
        product,
        quantity: 2,
      });
      cart.add({
        product: product2,
        quantity: 1,
      });

      cart.remove(product);

      expect(cart.getTotal()).toEqual(55060);
    });
  });

  describe("checkout()", () => {
    it("should return an object with the total price and the list of product item", () => {
      cart.add({
        product,
        quantity: 2,
      });
      cart.add({
        product: product2,
        quantity: 2,
      });

      expect(cart.checkout()).toMatchSnapshot();
    });

    it("should return an object with the total price and the list of product item when sumary() is called", () => {
      cart.add({
        product,
        quantity: 3,
      });
      cart.add({
        product: product2,
        quantity: 3,
      });

      expect(cart.sumary()).toMatchSnapshot();
      expect(cart.getTotal()).toBeGreaterThan(0);
    });

    it("should return getTotal equals 0 after checkout() to have been called", () => {
      cart.add({
        product,
        quantity: 2,
      });
      cart.checkout();

      expect(cart.getTotal()).toEqual(0);
    });
  });
});
