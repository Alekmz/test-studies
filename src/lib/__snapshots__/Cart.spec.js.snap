// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart test checkout() should return an object with the total price and the list of product item 1`] = `
{
  "items": [
    {
      "product": {
        "price": 45054,
        "title": "Nike shoes",
      },
      "quantity": 2,
    },
    {
      "product": {
        "price": 55060,
        "title": "Adidas shoes",
      },
      "quantity": 2,
    },
  ],
  "total": 200228,
}
`;

exports[`Cart test checkout() should return an object with the total price and the list of product item when sumary() is called 1`] = `
{
  "items": [
    {
      "product": {
        "price": 45054,
        "title": "Nike shoes",
      },
      "quantity": 3,
    },
    {
      "product": {
        "price": 55060,
        "title": "Adidas shoes",
      },
      "quantity": 3,
    },
  ],
  "total": 300342,
}
`;
